### PR TITLE
Fix right panel soft crashes chat rooms

### PR DESCRIPTION
**Pull request by toger5**
Fixes: https://github.com/vector-im/element-web/issues/20433

The right panel cannot be rendered if mandatory data is missing.
With this PR right panel cards, where required data is missing get removed prior to rendering the right panel.

This results in situations where:
A user views a specific member panel in the right panel
The user reloads element (page refresh) 
With the right panel refactor, it would be expected for the right panel to store its state and recover: displaying the same user.
Now the user panel will not be shown anymore and the right panel will show the user list after reload.
(This does not always happen. only if the local caching drops the user. I cannot exactly tell why that would happen. While trying to reproduce the issue I was always able to recover the user panel. But there seem to be situations in which the local caching does drop the user and thread events. apperantly.)

---
Here's what your changelog entry will look like:

## 🐛 Bug Fixes
 * Fix right panel soft crashes chat rooms ([\#7479](https://github.com/matrix-org/matrix-react-sdk/pull/7479)). Fixes vector-im/element-web#20433. Contributed by @toger5.

Preview: https://61d860a718c38e28d927a0ba--matrix-react-sdk.netlify.app
⚠️ Do you trust the author of this PR? Maybe this build will steal your keys or give you malware. Exercise caution. Use test accounts.

### CODE REVIEW

File: src/stores/right-panel/RightPanelStore.ts [hunk 1]
```diff
@@ -230,11 +230,19 @@ export default class RightPanelStore extends ReadyWatchingStore {
     }
 
     private emitAndUpdateSettings() {
+        if (this.global?.history) {
+            this.global.history = this.global?.history.filter((card)=>this.isCardStateValid(card));
```
Comment (t3chguy) [inline src/stores/right-panel/RightPanelStore.ts:234]:
```suggestion
            this.global.history = this.global?.history.filter((card) => this.isCardStateValid(card));
```

File: src/stores/right-panel/RightPanelStore.ts [hunk 2]
```diff
@@ -230,11 +230,19 @@ export default class RightPanelStore extends ReadyWatchingStore {
...
+        }
         const storePanelGlobal = convertToStorePanel(this.global);
         SettingsStore.setValue("RightPanel.phasesGlobal", null, SettingLevel.DEVICE, storePanelGlobal);
 
         if (!!this.viewedRoomId) {
-            const storePanelThisRoom = convertToStorePanel(this.byRoom[this.viewedRoomId]);
+            const panelThisRoom = this.byRoom[this.viewedRoomId];
+            if (panelThisRoom?.history) {
+                // remove all right panel cards that would fail to render (result in a soft crash)
+                panelThisRoom.history = panelThisRoom?.history.filter((card)=>this.isCardStateValid(card));
```
Comment (t3chguy) [inline src/stores/right-panel/RightPanelStore.ts:243]:
```suggestion
                panelThisRoom.history = panelThisRoom?.history.filter((card) => this.isCardStateValid(card));
```

File: src/stores/right-panel/RightPanelStore.ts [hunk 3]
```diff
@@ -230,11 +230,19 @@ export default class RightPanelStore extends ReadyWatchingStore {
     }
 
     private emitAndUpdateSettings() {
+        if (this.global?.history) {
```
Comment (t3chguy) [inline src/stores/right-panel/RightPanelStore.ts:233]: perhaps this filter stage can be a private method you call e.g `this.filterValidHistory()` or similar as the same if/filter is duplicated

File: src/stores/right-panel/RightPanelStore.ts [hunk 4]
```diff
@@ -245,6 +253,52 @@ export default class RightPanelStore extends ReadyWatchingStore {
...
+            case RightPanelPhases.ThreadView:
+                if ( !card.state.threadHeadEvent) {
+                    console.warn("removed card from right panel because of missing threadHeadEvent in card state");
+                }
+                return !!card.state.threadHeadEvent;
+            case RightPanelPhases.RoomMemberInfo:
+            case RightPanelPhases.SpaceMemberInfo:
+            case RightPanelPhases.EncryptionPanel:
+            case RightPanelPhases.GroupMemberInfo:
+                if ( !card.state.member) {
```
Comment (t3chguy) [inline src/stores/right-panel/RightPanelStore.ts:273]:
```suggestion
                if (!card.state.member) {
```

File: src/stores/right-panel/RightPanelStore.ts [hunk 5]
```diff
@@ -245,6 +253,52 @@ export default class RightPanelStore extends ReadyWatchingStore {
...
+                    console.warn("removed card from right panel because of missing member in card state");
+                }
+                return !!card.state.member;
+            case RightPanelPhases.SpaceMemberList:
+                if ( !card.state.spaceId) {
```
Comment (t3chguy) [inline src/stores/right-panel/RightPanelStore.ts:278]:
```suggestion
                if (!card.state.spaceId) {
```

File: src/stores/right-panel/RightPanelStore.ts [hunk 6]
```diff
@@ -245,6 +253,52 @@ export default class RightPanelStore extends ReadyWatchingStore {
...
+                    console.warn("removed card from right panel because of missing spaceId in card state");
+                }
+                return !!card.state.spaceId;
+            case RightPanelPhases.Room3pidMemberInfo:
+            case RightPanelPhases.Space3pidMemberInfo:
+                if ( !card.state.memberInfoEvent) {
```
Comment (t3chguy) [inline src/stores/right-panel/RightPanelStore.ts:284]:
```suggestion
                if (!card.state.memberInfoEvent) {
```

File: src/stores/right-panel/RightPanelStore.ts [hunk 7]
```diff
@@ -245,6 +253,52 @@ export default class RightPanelStore extends ReadyWatchingStore {
...
+                    console.warn("removed card from right panel because of missing memberInfoEvent in card state");
+                }
+                return !!card.state.memberInfoEvent;
+            case RightPanelPhases.GroupRoomInfo:
+                if ( !card.state.groupRoomId) {
```
Comment (t3chguy) [inline src/stores/right-panel/RightPanelStore.ts:289]:
```suggestion
                if (!card.state.groupRoomId) {
```

File: src/stores/right-panel/RightPanelStore.ts [hunk 8]
```diff
@@ -245,6 +253,52 @@ export default class RightPanelStore extends ReadyWatchingStore {
...
+                    console.warn("removed card from right panel because of missing groupRoomId in card state");
+                }
+                return !!card.state.groupRoomId;
+            case RightPanelPhases.Widget:
+                if ( !card.state.widgetId) {
```
Comment (t3chguy) [inline src/stores/right-panel/RightPanelStore.ts:294]:
```suggestion
                if (!card.state.widgetId) {
```

**Review by t3chguy — CHANGES_REQUESTED** [review]

File: src/stores/right-panel/RightPanelStore.ts (hunk 2 above)
Comment (toger5) [reply src/stores/right-panel/RightPanelStore.ts:243]: Can this be made a style rule? I do this mistake all the time (sorry)

Comment (t3chguy) [reply src/stores/right-panel/RightPanelStore.ts]: Good idea, feel free to add it to the eslint plugin https://github.com/matrix-org/eslint-plugin-matrix-org

**Review by t3chguy — APPROVED** [review]
Looks sane, thanks

**Comment by toger5** [comment]
> Looks sane, thanks

just wanted to request a review! thanks for beeing so quick!

**Comment by toger5** [comment]
Lets hope it fixes the issues. I could not reproduce it with the user panel. I could however reproduce it with the thread panel, so its hopefully fine.

**Comment by 33Fraise33** [comment]
Thanks for the fix, I will report back asap. The weird thing was that even on a "clear cache and reload" it did not resolve.

**Comment by toger5** [comment]
It is stored in the local site storage. If you don't need to log in again, it will also keep the right panel state (I think)

**Pull request merged**